### DISCUSSION TITLE
Can flipt support feature flags for a white label platform?

### DISCUSSION THREAD

**Discussion by prakashraman** [flipt-io/flipt#1885] [General]
I am building a white label application platform. Where features of my application screens need to be managed by a feature flag management tool (like flipt).

Therefore, for each customer, I would want to configure the flags as per their requirement. 

I understand that Flipt has a concept of "environments" which usually is intended for dev, staging and production. Can I use it to support customer specific flags? And if so, would there is a limitation on the number of environments I could create?

**Comment by markphelps** [comment]
Hi @prakashraman ! Great question! I Flipt could definitely be used for your use case. In fact, that's one of the reasons we decided to use the term [namespaces](https://www.flipt.io/docs/concepts#namespaces) instead of 'environments' like other feature flag platforms. While the avg team might use namespaces to separate their flags/segments by environment, you can use them however you like. 

There are no limitations on how many flags/segments/etc you can create, nor are there limitations on the number of namespaces you can create! 

The only thing I can think of is that in the UI we don't currently support pagination for the namespaces list view (where you create/edit namespaces). I created an internal issue for us to add this though, it won't be too difficult to implement.

Would love for you to give us any feedback on how Flipt works for you in your white-label application and any suggestions on how we can make it work better!

**Comment by prakashraman** [comment]
Hi Mark,

Thanks for the quick response! Really appreciate it.

Ah, I might have not explained one detail properly. I’d want the same set
of flags and segments available to all customers, but for each customer I’d
want to set different values for the flags. And I would have at least 10
customers.

Therefore I thought using “environments” would be the closest way.

P.s: pagination is not problem at all 😃

What do you think?

**Comment by prakashraman** [comment]
Ah! I just re read your message and read the docs again.

It does look like “namespaces” act as environments indeed.

Just to confirm: the same set of flags will be available for each namespace
right?

**Reply by GeorgeMac** [reply]
Hey @prakashraman 👋 

No, im afraid each namespace has unique content, unless you do some work to keep the definitions in sync.

To clarify, the experience you want is to centrally manage the definitions / names of flags and adjust their configuration / returned state on a per user or target basis?

Will you be managing which customers have which flags enabled?
Or will your customers be changing that themselves?

If you're the one managing which flags are enabled and for whom, then you could achieve this with Segments.
Segments let you identify any actor or group in the system (e.g. users) and then alter the state of flags at evaluation time based on who matches which segment.

You could make a segment per user and then create a rule to map that user to a flags particular evaluation result (either a particular variant or with the new boolean flag simply either `true` or `false`.

**Reply by prakashraman** [reply]
> No, im afraid each namespace has unique content, unless you do some work to keep the definitions in sync.
I see, ok.

> To clarify, the experience you want is to centrally manage the definitions / names of flags and adjust their configuration / returned state on a per user or target basis?
Exactly, yes on a target/customer basis

> Or will your customers be changing that themselves?
No, we will manage it internally

> If you're the one managing which flags are enabled and for whom, then you could achieve this with Segments.
> Segments let you identify any actor or group in the system (e.g. users) and then alter the state of flags at evaluation time based on who matches which segment.
This could work, thanks!

Unrelated. Does "flags" have types? (i.e boolean flags, string flags, array of strings flags, enum flags) I was unable to find this specifically in the documentation. 

Thanks George.

**Reply by GeorgeMac** [reply]
Historically, we just had one flag type and that is true of Flipt latest version.
That type is what we call "variant" type flag.

However, we are releasing a release candidate this week of Flipt v1.24 which contains a new type "boolean".

By next week we will be aiming for a more general release and to have some good documentation around the differences between these types.

However, variant type flags have multiple variant keys (kind of like a string type flag) and targetting rules decide which variant is returned at evaluation time and boolean type is as you would expect, the value is either true or false and rules dictate that at evaluate time.
